lsコマンドを作る1の提出物です

rira100000000 opened the pull request:
# 内容
lsコマンドを作る1の初回提出です。

# 背景
プラクティス"lsコマンドを作る"の課題
[lsコマンドを作る1 \| FBC](https://bootcamp.fjord.jp/practices/221)

rira100000000 committed:
カレントディレクトリの情報取得機能を追加

カレントディレクトリ内のファイル名、ディレクトリ名を取得するget_file_listを追加
get_file_listが取得した要素の数が正しいかテストするtest_get_file_listを追加

rira100000000 committed:
テスト用ファイルを追加

テストに使用する空ファイルとディレクトリを追加
それらを生成するmake_test_file.rbを追加

rira100000000 committed:
RuboCop対応

RuboCopの指摘部分修正
定数SPACE_FOR_COLUMNS追加

rira100000000 committed:
テストの追加とリファクタリング

test_make_disp_linesの追加
test_print_file_listの処理の一部をadd_space_for_lineとメイン処理に分離
test_print_file_listの名前をmake_disp_linesに変更

rira100000000 committed:
RuboCop対応

rira100000000 committed:
コマンドライン引数に対応

コマンドラインから引数を付けて実行できるように修正
引数をオプションとpathに分けるsplit_option_or_pathを追加
pathが複数ある場合、それぞれのpathのファイル一覧を表示するように修正

rira100000000 committed:
パスが複数指定された場合のテストを追加

パスが複数の場合のテストを追加
メインの処理をmake_disp_strに移動
列数が指定と異なる場合があったので修正

rira100000000 committed:
RuboCop対応

rira100000000 committed:
ファイル名の全角表記に対応

ファイル名が全角の時、表示が崩れないように修正

rira100000000 committed:
RuboCop対応

rira100000000 committed:
引数でファイルを指定された場合に対応

コマンドライン引数でファイル名を指定された場合の処理を追加
同、テストを追加

rira100000000 committed:
ファイル名が重複して表示される不具合を修正

ファイル名が重複して表示される不具合を修正
テストの修正

rira100000000 committed:
RuboCop対応

make_disp_strの処理の一部をanalys_directory_pathsに分離
make_disp_strの処理の一部をanalys_file_pathsに分離

rira100000000 committed:
メソッド名の変更

何をしているのかわからないメソッド名を、より具体的なものに変更した

yoshitsugu commented on 04.ls/test/ls_test.rb:
```diff
@@ -0,0 +1,55 @@
+require 'minitest/autorun'
+require_relative '../lib/ls'
+
+class ListTest < Minitest::Test
```
テストしっかり書いててすごいですね :+1:

yoshitsugu commented on 04.ls/lib/ls.rb:
```diff
@@ -0,0 +1,111 @@
+SPACE_FOR_COLUMNS = 2
+def glob_file_list(path)
+  Dir.glob('*', base: path)
+end
```
ここあまりメソッド化するメリットがないかもです。
メソッドとして抽象化する利点としては、

- 処理を意味の単位で一段抽象的な処理としてまとめて、再利用できる
- タイプ数が減らせる
- メソッドの中身の処理が変わってもメソッドの入出力が同じならメソッドを呼び出す側は変更が不要

などがありますが、逆にデメリットとして、
  
- 意味の単位でまとまっていないと、用途が不明瞭で再利用ができない
- 処理を細かく追いたいときにはメソッドの中身を見ないといけないので、コードを読むときに視線が無駄に上下してしまう
- アルゴリズムが隠蔽されてしまい、計算効率がわかりにくい

というのがあり、いつでもメソッドにすればOK、というわけではないので注意してください！

yoshitsugu commented on 04.ls/lib/ls.rb:
```diff
@@ -0,0 +1,111 @@
+  end
+end
+
+def analys_directory_paths(paths)
```
`analyse` と `e` が入るかなと思います。

yoshitsugu commented on 04.ls/lib/ls.rb:
```diff
@@ -0,0 +1,111 @@
+  ARGV
+end
+
+def make_disp_list(argv)
```
`disp` は `display` とかですかね。メソッド名や変数名は多少長くなっても略さないで明示的に意味がわかるほうを重視することが多いです。
なぜかというと、以下あたりの理由です。

- 一般的にコードは書かれるより読まれる時間のほうが長いため、読みやすいことに価値があること
- IDEなどの支援があれば長い変数名なども補完でき、短かくするメリットがあまりないこと

yoshitsugu commented on 04.ls/lib/ls.rb:
```diff
@@ -0,0 +1,111 @@
+def calc_file_name_size(file_name)
+  count = 0
+  file_name.each_char do |char|
+    count += char.ascii_only? ? 1 : 2
```
おお、細かい判定までやっていますね。
ちなみに、豆知識レベルですが、ascii_only? ではなくても表示が半角1文字分の文字を作ることはできて、
ウムラウト記号とかがそれになります。

```
irb(main)[02:0]> "ő".ascii_only?
=> false
```

とはいえこの辺まで考えだすとキリがないので、今回はこのくらいでOKだと思います！

yoshitsugu commented on 04.ls/lib/ls.rb:
```diff
@@ -0,0 +1,111 @@
+end
+
+def calc_file_name_size(file_name)
+  count = 0
+  file_name.each_char do |char|
+    count += char.ascii_only? ? 1 : 2
+  end
+  count
```
sumのブロックとして渡す方法もありますね。

```ruby
  file_name.each_char.sum do |char|
    char.ascii_only? ? 1 : 2
  end
```

基本的に、

```ruby
array2 = []
array1.each do |item1|
   array2 << xxx(item1)
end
array2
```

みたいな形を見たら、


```ruby
array1.map { |item1| xxx(item1) }
```

と書き換えるなど、配列 -> 配列の操作はいくつかありますので、 `map` 、 `select` 、 `sum` 、 `max` 、 できれば `inject` あたりも使いこなせるとコードをもっと簡潔に書ける場合が増えるかなーと思います。

yoshitsugu commented on 04.ls/lib/ls.rb:
```diff
@@ -0,0 +1,111 @@
+  lines.each do |file_names|
+    disp_line = ''
+    file_names.each_with_index do |file_name, i|
+      disp_line += file_name + ' ' * (max_file_names[i] - calc_file_name_size(file_name) + SPACE_FOR_COLUMNS)
```
文字列連結はstring interpolation(式展開)を使うことが多いです。
なぜかというと `+` だともし `nil + " "` になったときにエラーになりますが、string interpolationだと `"#{nil} "` も特にエラーになりません。
もちろんエラーになってほしい場面も多々あるのでケースバイケースではありますが、string interpolationで足りる場面が多いかなと思います。

yoshitsugu commented on 04.ls/lib/ls.rb:
```diff
@@ -0,0 +1,111 @@
+end
+
+def analys_file_paths(paths)
+  file_list = []
+  paths.each { |path| file_list << path if File::Stat.new(path).file? }
+  file_list
```
ここも `map` でよさそうですね。

yoshitsugu reviewed (commented):
コメントしました！

rira100000000 committed:
テストのバグ修正

カレントディレクトリによってテストが通らなくなるバグを修正

rira100000000 committed:
メソッドgrob_file_listを削除

可読性を高めるため、不要なメソッドを削除しました

rira100000000 committed:
メソッド名の修正

メソッド名で使用されるanalys→analyseに修正しました
メソッド名で使用されるdisp→displayに変更しました

rira100000000 committed:
リファクタリング

sumメソッドを使えるところでは使用するように修正
eachで配列処理していたのをselectに修正

rira100000000 committed:
文字列連結で式展開を使用するよう修正

文字列連結に+を使っていたところを式展開を使用するよう修正

rira100000000 committed:
RuboCop対応

yoshitsugu commented on 04.ls/lib/ls.rb:
```diff
@@ -0,0 +1,104 @@
+  end
+end
+
+def split_option_or_path(argv)
+  options = {}
+  paths = []
+  argv.each do |str|
+    if ['-a', '-r', '-l'].include?(str)
+      options[str] = true
+    else
+      paths << str
+    end
+  end
+  [options, paths]
+end
```
わけたいのであれば `parse!` を使う手もありますよ
https://docs.ruby-lang.org/ja/latest/method/OptionParser/i/parse=21.html

yoshitsugu commented on 04.ls/lib/ls.rb:
```diff
@@ -0,0 +1,104 @@
+def add_space_for_line(lines, max_file_names)
+  result = []
+  lines.each do |file_names|
+    display_line = ''
+    file_names.each_with_index do |file_name, i|
+      display_line = "#{display_line}#{file_name}#{' ' * (max_file_names[i] - calc_file_name_size(file_name) + SPACE_FOR_COLUMNS)}"
+    end
```
毎回新しい文字列を作るのはちょっと効率悪そうな気がしますね。
`map` してStringの配列を一旦作ってから `join` するとか、`String#<<` を使うとかですかね。

https://docs.ruby-lang.org/ja/latest/method/String/i/=3c=3c.html

yoshitsugu commented on 04.ls/lib/ls.rb:
```diff
@@ -0,0 +1,104 @@
+end
+
+def add_space_for_line(lines, max_file_names)
+  result = []
+  lines.each do |file_names|
+    display_line = ''
+    file_names.each_with_index do |file_name, i|
+      display_line = "#{display_line}#{file_name}#{' ' * (max_file_names[i] - calc_file_name_size(file_name) + SPACE_FOR_COLUMNS)}"
+    end
+    result << display_line
+  end
+  result
```
ここの`each` も `map` にかえられそうですね。

yoshitsugu reviewed (commented):
コメントしました！

rira100000000 committed:
split_option_or_pathの削除

parse!メソッドを使うことで簡潔に書けるため、split_option_or_pathを削除しました
付随してparse_optionの戻り値がpathsから[paths,options]に変更されました

rira100000000 committed:
テストの追加

オプション、パスの指定がない時のテストを追加

rira100000000 committed:
文字列連結をstring#<<に変更

速度効率を考え文字列生成していた部分をstring#<<に変更しました

rira100000000 committed:
eachをmapに変更

mapで書ける処理はmapで書くよう修正

rira100000000 committed:
rubocop対応

rubocop指摘部分を修正

yoshitsugu commented on 04.ls/test/ls_test.rb:
```diff
@@ -0,0 +1,70 @@
+  end
+
+  # ファイルとパスが複数指定された場合
+  def test_make_display_list_with_file_name_and_paths
+    result = `ruby #{__dir__}/ls_args_file_name_and_paths_test.rb \
+    #{__dir__}/test_data/01_file #{__dir__}/test_data/02dir \
+    #{__dir__}/test_data/03_file #{__dir__}/test_data/04dir`
+    assert result.include?('0 failures, 0 errors, 0 skips')
```
手元でこいつが失敗しているっぽいんですが、そちらではうまくいっていますか？

yoshitsugu reviewed (commented):
ほぼよさそうです！一点だけコメントしました。

yoshitsugu reviewed (approved):
一旦OKです。
